text/0000-try-trait.md

```diff
@@ -0,0 +1,553 @@
+`Carrier` was implemented separately for each type (e.g., `Result` and
+`Option`). It allowed any kind of carrier to be converted to any other
+kind of carrier, which fails to preserve the "semantic equivalent"
+property.  It would also interact poorly with type inference, as
+discussed in the "Detailed Design" section.
+
+### Traits implemented over higher-kinded types
+
+The "semantic equivalent" property might suggest that the `Carrier`
+trait ought to be defined over higher-kinded types (or generic
+associated types) in some form. The most obvious downside of such a
+design is that Rust does not offer higher-kinded types nor anything
+equivalent to them today, and hence we would have to block on that
+design effort. But it also turns out that HKT is
+[not a particularly good fit for the problem](https://github.com/rust-lang/rust/pull/35056#issuecomment-240129923). To
+start, consider what "kind" the `Self` parameter on the `Try` trait
+would have to have.  If we were to implement `Try` on `Option`, it
+would presumably then have kind `type -> type`, but we also wish to
+implement `Try` on `Result`, which has kind `type -> type ->
+type`. There has even been talk of implementing `Try` for simple types
+like `bool`, which simply have kind `type`. More generally, the
+problems encountered are quite similar to the problems that
+[Simon Peyton-Jones describes in attempting to model collections using HKT](https://github.com/rust-lang/rust/pull/35056#issuecomment-240129923):
+we wish the `Try` trait to be implemented in a great number of
+scenarios.  Some of them, like converting `Result<T,E>` to
+`Result<U,F>`, allow for the type of the success value and the error
+value to both be changed, though not arbitrarily (subject to the
+`From` trait, in particular).  Others, like converting `Option<T>` to
+`Option<U>`, allow only the type of the success value to change,
+whereas others (like converting `bool` to `bool`) do not allow either
+type to change.
+
+### What to name the trait
+
+A number of names have been proposed for this trait. The original name
+was `Carrier`, as the implementing type was the "carrier" for an error
+value. A proposed alternative was `QuestionMark`, named after the
+operator `?`. However, the general consensus seemed to be that since
+Rust operator overloading traits tend to be named after the
+*operation* that the operator performed (e.g., `Add` and not `Plus`,
+`Deref` and not `Star` or `Asterix`), it was more appropriate to name
+the trait `Try`, which seems to be the best name for the operation in
+question.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+**We need to resolve the interactions with type inference.** It is
+important that expressions like `vec.iter().map(|e| ...).collect()?`
+are able to infer the type of `collect()` from context. This may
+require small tweaks to the `Try` trait, though the author considers
+that unlikely.
+
```

Comment:
I think that `try!()` forces collect to pick the `Result` impl is merely a cute trick — the usual way to use `.collect()` is to specify which type you are collecting to, so specifying that inside `.collect::<Result<Vec<_>, _>>()?` seems frankly normal.

I am just pointing out that there is a problem with type inference and error conversion, and I am not proposing that it should be removed. But it would be a lot more useful if no-ops (no conversion) was a fallback mode.